### PR TITLE
(PUP-5595) Do not retrieve generated resources until they are applied

### PR DESCRIPTION
**Pull request by branan**
PUP-1963 changed resource generation such that resources were added to
the *catalog*, not just the final relationship graph. This required
the Type instances returned by the `generate` method to be converted
to Resource instances.

Unfortunately, the `to_resource` method calls `retrieve`, presumably
for the sake of the RAL. This would cause resources to be retrieved at
generation time, rather than at application time.

Instead of using `to_resource`, the additional resource generator now
constructs a Resource instance using its own logic, including only the
parameters that were used when the Type instance was created.

### CODE REVIEW

File: lib/puppet/transaction/additional_resource_generator.rb [hunk 1]
```diff
@@ -170,8 +170,11 @@ def add_generated_directed_dependency(parent, child, label=nil)
     }
 
     if not edge_exists
+      parameters = target.original_parameters
+      resource = ::Puppet::Resource.new(target.class, target.title, :parameters => parameters)
```
Comment (branan) [inline lib/puppet/transaction/additional_resource_generator.rb]: I was having issues with namespacing here, and being totally explicit seemed the best way to move forward

**Comment by reidmv** [comment]
I applied this patch locally and verified that it resolves PUP-5595.

File: lib/puppet/transaction/additional_resource_generator.rb (hunk 1 above)
Comment (kylog) [inline lib/puppet/transaction/additional_resource_generator.rb]: Style (I think) nit, but is there any reason not to forgo the local and do something like:

```
      resource = ::Puppet::Resource.new(target.class, target.title, 
                                        :parameters => target.original_parameters)
```

**Comment by kylog** [comment]
Is there anything (not insane) we can do at spec test level to enforce this doesn't regress?

File: lib/puppet/transaction/additional_resource_generator.rb (hunk 1 above)
Comment (branan) [inline lib/puppet/transaction/additional_resource_generator.rb]: I had split it for debug prints, and never put it back together. I don't have strong opinions about the format here, I'm hapy to change it.

**Comment by branan** [comment]
@kylog I can add a spec test that generate doesn't call retrieve, but that feels very "bug-fix-specific-unit-test" to me. I feel like these old low-level puppet classes are a mess, and any given unit test probably won't actually prevent real regressions (which will probably happen in another code path)

I should definitely add a comment as to why to_resource isn't being used, though

**Comment by kylog** [comment]
@branan :+1: on adding an explanatory comment rather than a test-the-implementation-not-the-interface spec test.

**Comment by branan** [comment]
@kylog updated with comment, and addressed your style nit

**Comment by kylog** [comment]
:+1: 

Will let travis go green.

**Comment by kylog** [comment]
Travis seems borked. Let's try Jenkins.

**Pull request merged**